### PR TITLE
Adds a guard statement to prevent NoMethodErrors

### PR DESCRIPTION
**Pull request by empty-codes**
Temporary workaround for Issue #31 by:
1. Adding a `next unless revisions` guard statement to prevent errors like NoMethodError: undefined method 'each' for nil:NilClass, which occurs if the code attempts to iterate over a nil value (i.e., if revisions is nil).
2. Creating a spec for this expected behaviour:
- If revisions is nil (or doesn't exist), the condition unless revisions becomes true, and the next statement skips the rest of the current loop iteration.
- If revisions is not nil (i.e., it exists and is an array or object), the loop continues to process the revisions.

### Before:
image

### After:
image

### CODE REVIEW

**Comment by ragesoss** [comment]
I recommend both adding a test for this behavior (one which would fail before the change), and also adding a comment explaining the possibility of an API response that has a page with no revisions.

**Comment by empty-codes** [comment]
> I recommend both adding a test for this behavior (one which would fail before the change), and also adding a comment explaining the possibility of an API response that has a page with no revisions.

@ragesoss  Done! I tried to make the test as simple as possible and the comments easy to understand. I have updated the description with screenshots of running the spec.

File: spec/wikidata/diff/api_spec.rb [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+          2266341148, 2266123442, 2266123459, 2266123479, 2266123502, 2266123529, 
+          2266123536, 2266123548, 2266123562, 2266123568, 2266341782, 2266123581, 
+          2266123596, 2266123602
+        ]
+      end
+  
+      it 'returns without raising an error when an API response has a page with no revisions.' do
+        expect {
+          Api.get_revision_contents(revision_ids)
+        }.not_to raise_error
```
Comment (ragesoss) [inline spec/wikidata/diff/api_spec.rb:36]: I recommend including some additional expectations about the size and shape of the return value from this API call (such as expecting there to be 48 entries rather than 50)

**Comment by ragesoss** [comment]
Thanks! I guess this will also require a minor code adjustment on the Dashboard side, since we no longer can expect one entry in the results for every rev_id we pass in to the WikidataDiffAnalyzer.

File: spec/wikidata/diff/api_spec.rb (hunk 1 above)
Comment (empty-codes) [reply spec/wikidata/diff/api_spec.rb:36]: I added these additional expectations to the spec:

```
it 'returns without raising an error and handles both cases where revisions are present or absent' do
        expect {
          @result = Api.get_revision_contents(revision_ids)
        }.not_to raise_error
  
        expect(@result).to be_a(Hash)

        expect(@result.size).to eq(48)

      end
```
        
But when i ran the spec, it failed since `result.size` was 30 instead of 48:
image


Now, result is what is returned by the `get_revision_contents` which is the `parsed_contents` which is a hash.
I was confused at first but apparently, each `revision_id` corresponds to a specific change made to a page in Wikidata so multiple `revision_ids` can be associated with a single page.
In a case like this where a page has no revisions, all the `revision_ids` associated with the page, do not appear in the response.

I tested this by adding statements to check the number of `pages`, `revisions` and `parsed_count`objects.

For the batch of 50 with the 2 pages having no revisions:
image

Then i tried random batches of 50 (all their pages had revisions) from the total revision_ids array gotten from the course (it's a whole lot) and they all had 50 entries as expected. Screenshots of the result of two of them:

random batch 1:
image

random batch 2:
image

I guess we can surmise that the missing 20 in the affected batch of 50 is shared between the 2 pages that unfortunately have no revisions.

When I change the expected result size to 30 `  expect(@result.size).to eq(30)`, the spec passes fine.
So in conclusion, are the additional expectaions of checking if the result is a hash and the result size enough? If so should I set the result size to 30?

Comment (ragesoss) [reply spec/wikidata/diff/api_spec.rb:36]: I'd prefer to have a test where the inputs and outputs are precisely documented. Maybe you can either find a dataset where the missing pages only have one revision each, or otherwise document which revisions correspond to which pages? I don't think we need to test with a full batch of 50. A batch of 3 to 5 would be fine.

**Comment by empty-codes** [comment]
@ragesoss , following your suggestion to  find a dataset where the missing pages only have one revision each, I found out I mis - debugged this whole issue entirely.

This is the full query sent for the problematic batch: https://www.wikidata.org/w/api.php?action=query&prop=revisions&revids=2266122608|2266122618|2266122626|2266122646|2266122666|2266122683|2266122709|2266122730|2266122739|2266122747|2266122763|2266122777|2266122783|2266122790|2266122808|2266122817|2266122829|2266122850|2266122880|2266122931|2266122949|2266122973|2266122994|2266123011|2266123017|2266123021|2266341034|2266123060|2266123123|2266123148|2266123175|2266123210|2266123270|2266123325|2266123373|2266123418|2266341148|2266123442|2266123459|2266123479|2266123502|2266123529|2266123536|2266123548|2266123562|2266123568|2266341782|2266123581|2266123596|2266123602&rvslots=main&rvprop=content|ids|comment&format=json

If you visit the link, the first thing you will notice is the presence of a 'rvcontinue'.
```
{"continue":{"rvcontinue":2266123210,"continue":"||"},"warnings":{"result":{"*":"This result was truncated because it would otherwise be larger than the limit of 12,582,912 bytes."}},
```
Initially i just logged the api response following the parameter structure I saw in api.rb, without actually checking the full thing, which was my oversight.

So this message indicates that the API response is truncated due to size limits imposed by the MediaWiki API. The API has a maximum response size limit of 12,582,912 bytes (approximately 12.5 MB), and when the response exceeds this size, it provides only a partial result.

This means the response  received does not contain all the requested revision data. Instead, it includes only a portion of the requested revisions. So even though we requested revisions for all the given revision IDs, only some of them are returned in this truncated response.

### To understand what was happening, I did more testing:

 I divided the batch of 50 into 2:

- the first batch of 25: https://www.wikidata.org/w/api.php?action=query&format=json&formatversion=2&prop=revisions&revids=2266122608|2266122618|2266122626|2266122646|2266122666|2266122683|2266122709|2266122730|2266122739|2266122747|2266122763|2266122777|2266122783|2266122790|2266122808|2266122817|2266122829|2266122850|2266122880|2266122931|2266122949|2266122973|2266122994|2266123011|2266123017&rvprop=ids%7Ccomment%7Ccontent&rvslots=main returns ```"batchcomplete":true,"``` and all 25 revisions are parsed
image

- now the second half: https://www.wikidata.org/w/api.php?action=query&format=json&formatversion=2&prop=revisions&revids=2266123021|2266341034|2266123060|2266123123|2266123148|2266123175|2266123210|2266123270|2266123325|2266123373|2266123418|2266341148|2266123442|2266123459|2266123479|2266123502|2266123529|2266123536|2266123548|2266123562|2266123568|2266341782|2266123581|2266123596|2266123602&rvprop=ids%7Ccomment%7Ccontent&rvslots=main returns ```{"continue":{"rvcontinue":2266123373,"continue":"||"},"warnings":{"result":{"warnings":"This result was truncated because it would otherwise be larger than the limit of 12,582,912 bytes."}}``` and only 8 revisions are parsed instead of the 25:
image

- So then i ran the spec only on the 17 revision ids that failed: https://www.wikidata.org/w/api.php?action=query&format=json&formatversion=2&prop=revisions&revids=2266123373|2266123418|2266341148|2266123442|2266123459|2266123479|2266123502|2266123529|2266123536|2266123548|2266123562|2266123568|2266341782|2266123581|2266123596|2266123602&rvprop=ids%7Ccomment%7Ccontent&rvslots=main, the response here returns "batchcomplete":true," and all 17 revisions are parsed including the so called 'have no revisions' page ids 188280 and 265881
image

I still wasn't convinced so I ran the spec using only 1 revision id each time from the batch of 17, and they all had revisions, were all parsed successfully and had no errors each time.

### So in conclusion, it is not that certain pages do not have revisions but rather the gem does not have logic to handle when responses get truncated and instead interprets it as the pages having no revisions.

This `api_spec.rb` spec calls the get_revision_contents() method directly, so I tried reducing the batch size from 50 to 25 here: https://github.com/WikiEducationFoundation/wikidata-diff-analyzer/blob/7480f31f15d8438bdfccb0214a00834c1f59bfa3/lib/wikidata-diff-analyzer.rb#L76 and calling the analyze method on the whole revision ids set for the course but that did not make any difference, because well it's a response limit issue, so while some batches correctly included 25 entries , some would include just 7, 13, 22 and so on.

I should have been way more thorough in debugging this from the beginning, my apologies😅.  Then again, this makes a pretty good lesson on why tests are so important.

**Comment by ragesoss** [comment]
Excellent work!

I guess now we're back to an open question on what to do about this. Can you figure out how the truncation works? If it simply leaves out the last revision that pushes it over the size limit, we could identify which revisions we received results for, and make a new request with just the remaining revisions. But if it can include partial data for a single revision, we need to either figure out which revision and try requesting that one by itself, or we need throw out the whole request and break it up into smaller batches every time it happens.

**Comment by empty-codes** [comment]
@ragesoss  Yes, I can investigate how the truncation works. In the meantime, my plan is to:

1. Update the comments to clearly explain what’s happening and note that this is a temporary workaround as suggested
2. For the test, should I still include expectations on the result size (e.g., specifying that for a batch of 25 revisions, only 8 will be processed), or would it be better to leave those expectations out entirely?
3. Modify the Dashboard code to handle what it expects returned ( I’ll create a PR under the existing issue there)
After that, I’ll begin working on a proper fix. Should I go ahead?

**Comment by ragesoss** [comment]
Hmm.... I guess since we're working with specific revisions, it's unlikely that the behavior of the query would change over time (unlike if we were using item ids, where new edits to that Wikidata item would increase the response size over time). However, I don't think any additional tests are necessary; for the current behavior, we know that it is not strictly correct, and not erroring is the part of the behavior that we know we want to keep. The result size is something that might change based on how we implement a for the truncation issue.

File: spec/wikidata/diff/api_spec.rb [hunk 2]
```diff
@@ -6,31 +6,31 @@
...
-          2266122783, 2266122790, 2266122808, 2266122817, 2266122829, 2266122850, 
-          2266122880, 2266122931, 2266122949, 2266122973, 2266122994, 2266123011, 
-          2266123017, 2266123021, 2266341034, 2266123060, 2266123123, 2266123148, 
-          2266123175, 2266123210, 2266123270, 2266123325, 2266123373, 2266123418, 
-          2266341148, 2266123442, 2266123459, 2266123479, 2266123502, 2266123529, 
-          2266123536, 2266123548, 2266123562, 2266123568, 2266341782, 2266123581, 
-          2266123596, 2266123602
+          # A batch of 25 revision IDs associated with a course (course_id: 10023).
+          # JSON: https://www.wikidata.org/w/api.php?action=query&prop=revisions&revids=2266123021|2266341034|2266123060|2266123123|2266123148|2266123175|2266123210|2266123270|2266123325|2266123373|2266123418|2266341148|2266123442|2266123459|2266123479|2266123502|2266123529|2266123536|2266123548|2266123562|2266123568|2266341782|2266123581|2266123596|2266123602&rvslots=main&rvprop=content|ids|comment&format=json
+          # This batch includes pages (pageid: 18820 and 265881) that appear to lack the 'revisions' property.
```
Comment (ragesoss) [inline spec/wikidata/diff/api_spec.rb:24]: I would be more specific about the response we are testing against, noting that the query is expected to pass the size limit and include a warning about truncation.

File: spec/wikidata/diff/api_spec.rb [hunk 3]
```diff
@@ -6,31 +6,31 @@
...
+          2266123021, 2266341034, 2266123060, 2266123123, 2266123148, 
+          2266123175, 2266123210, 2266123270, 2266123325, 2266123373, 
+          2266123418, 2266341148, 2266123442, 2266123459, 2266123479, 
+          2266123502, 2266123529, 2266123536, 2266123548, 2266123562, 
+          2266123568, 2266341782, 2266123581, 2266123596, 2266123602
         ]
       end
   
-      it 'returns without raising an error when an API response has a page with no revisions.' do
+      it 'returns without raising an error and handles both cases where revisions are present or absent' do
```
Comment (ragesoss) [inline spec/wikidata/diff/api_spec.rb:33]: Here, perhaps we should also include an expectation about the truncation warning. If the behavior of this API with regards to large queries changes, that is something that we probably want to break the test.

File: spec/wikidata/diff/api_spec.rb (hunk 2 above)
Comment (empty-codes) [reply spec/wikidata/diff/api_spec.rb:24]: How is this: 
```  
# A batch of 25 revision IDs associated with course_id: 10023.
# JSON: https://www.wikidata.org/w/api.php?action=query&prop=revisions&revids=2266123021|2266341034|2266123060|2266123123|2266123148|2266123175|2266123210|2266123270|2266123325|2266123373|2266123418|2266341148|2266123442|2266123459|2266123479|2266123502|2266123529|2266123536|2266123548|2266123562|2266123568|2266341782|2266123581|2266123596|2266123602&rvslots=main&rvprop=content|ids|comment&format=json
# This query exceeds the MediaWiki API's response size limit of 12,582,912 bytes, resulting in a truncated response.
# The response includes a 'warnings' property and only a subset of the requested revisions.
# Warnings: "This result was truncated because it would otherwise be larger than the limit of 12,582,912 bytes."
# This truncation causes pages (pageid: 18820 and 265881) to lack the 'revisions' property.
```

File: spec/wikidata/diff/api_spec.rb (hunk 3 above)
Comment (empty-codes) [reply spec/wikidata/diff/api_spec.rb:33]:
```ruby
 it 'returns a truncation warning if query exceeds response size limit' do
        API_URL = 'https://www.wikidata.org/w/api.php'
        client = MediawikiApi::Client.new(API_URL)
        
        response = Api.fetch_revision_data(client, revision_ids)

        expect(response.warnings)

        expect(response.warnings).to include("This result was truncated because it would otherwise be larger than the limit of 12,582,912 bytes.")
      end
```
I spent a while figuring out how MediawikiApi::Client works, but this works

**Comment by ragesoss** [comment]
Looks good. Fix the linting error and I think it will be okay to merge.

**Comment by empty-codes** [comment]
> Excellent work!
> 
> I guess now we're back to an open question on what to do about this. Can you figure out how the truncation works? If it simply leaves out the last revision that pushes it over the size limit, we could identify which revisions we received results for, and make a new request with just the remaining revisions. But if it can include partial data for a single revision, we need to either figure out which revision and try requesting that one by itself, or we need throw out the whole request and break it up into smaller batches every time it happens.

@ragesoss  I have finally solved this 🥳.
It simply leaves out the last revision that pushes it over the size limit since it uses continue: https://www.mediawiki.org/wiki/API:Continue so I was able to use continue to get the remaining responses.

I created a new branch and PR for a cleaner commit history: https://github.com/WikiEducationFoundation/wikidata-diff-analyzer/pull/33
